include/hw_unique_key.h

```diff
@@ -20,15 +20,119 @@
 extern "C" {
 #endif
 
+#include <stdbool.h>
+
+#ifdef __ZEPHYR__
+#include <devicetree.h>
+#define HUK_HAS_KMU DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_kmu)
+#define HUK_HAS_CC310 DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_cc310)
+#define HUK_HAS_CC312 DT_HAS_COMPAT_STATUS_OKAY(nordic_nrf_cc312)
+#else
+#define HUK_HAS_KMU 1
+#define HUK_HAS_CC310 defined(NRF9160_XXAA)
+#define HUK_HAS_CC312 defined(NRF5340_XXAA_APPLICATION)
+
+#endif /* __ZEPHYR__ */
+
+#if HUK_HAS_CC310
+#define HUK_SIZE_WORDS 4
+#elif HUK_HAS_CC312
+#define HUK_SIZE_WORDS 8
+#endif
+#define HUK_SIZE_BYTES (HUK_SIZE_WORDS * 4)
+
+#define ERR_HUK_MISSING 0x15500
+
+
+/* The available slots. KDR is always available, while the MKEK and MEXT
+ * keys are only stored when there is a KMU, since without a key, the key
+ * store must be locked after booting, and the KDR is the only key that can
+ * live in the CC HW for the entire boot cycle of the device.
+ */
+#if HUK_HAS_CC310
+#define HUK_KEYSLOT_KDR  0 /* Device Root Key */
+#endif
+#if HUK_HAS_KMU
+#define HUK_KEYSLOT_MKEK 2 /* Master Key Encryption Key */
+#define HUK_KEYSLOT_MEXT 4 /* Master External Storage Encryption Key */
+#endif
+
+/* The available slots as an array that can be iterated over. */
+static const uint32_t slots[] = {
+#if HUK_HAS_CC310
+	HUK_KEYSLOT_KDR,
+#endif
+#if HUK_HAS_KMU
+	HUK_KEYSLOT_MKEK,
+	HUK_KEYSLOT_MEXT,
+#endif
+};
+
+#define KMU_SELECT_SLOT(KEYSLOT) ((KEYSLOT) + 1) /* NRF_KMU KEYSLOT are 1-indexed. */
+
+/**
+ * @brief Write a Hardware Unique Key to the KMU. This can only be done once
+ *       (until a mass erase happens). Panic on failure.
+ *
+ * @param[in] kmu_keyslot  The keyslot to write to, see HUK_KEYSLOT_*.
+ * @param[in] key          The key to write. Must be @ref HUK_SIZE_BYTES bytes long.
+ */
+void hw_unique_key_write(uint32_t kmu_keyslot, const uint8_t *key);
+
+#ifdef CONFIG_HW_UNIQUE_KEY_RANDOM
 /**
- * @brief This function loads the Hardware Unique Key (HUK) into the KDR
- *        registers of the Cryptocell. It also locks the flash page
- *        which contains the key material from reading and writing
- *        until the next reboot.
+ * @brief Read a random number from @ref mbedtls_hardware_poll and write it
+ *        with @ref hw_unique_key_write. Panic on failure.
+ */
+void hw_unique_key_write_random(void);
+#endif /* CONFIG_HW_UNIQUE_KEY_RANDOM */
+
+/**
+ * @brief Check whether a Hardware Unique Key has been written to the KMU.
+ *
+ * @param[in] kmu_keyslot  The keyslot to check, see HUK_KEYSLOT_*.
+ *
+ * @return true if a HUK has been written, false otherwise.
+ */
+bool hw_unique_key_is_written(uint32_t kmu_keyslot);
+
+/**
+ * @brief Check whether all Hardware Unique Keys are written to the KMU.
+ *
+ * @return true if all HUKs are written, false if one or more HUK is unwritten.
+ */
+bool hw_unique_key_are_any_written(void);
+
+/**
+ * @brief Load the Hardware Unique Key (HUK) into the KDR registers of the
```

Comment:
It is not possible for this to fail in any way? Is it possible that no HUK is generated/loaded?

Reply:
These functions all panic on failure. I made the docs more clear.